### PR TITLE
feat: Add branch protection rules changes 

### LINKED ISSUE #26877: Add branch protection feature for the EE users
We need to achieve the following:
1. Add API to protected and unprotect a branch
2. Get list of branches with their protection status
3. Disable commit, delete, pull in protected branches
4. Add fallback implementation for the Free users

### PR DESCRIPTION
**Pull request by AnaghHegde**
## Description
This PR adds check for handling the branch protection rules before performing the git commit, pull and delete ops. This is fall back for CE behaviour. This is paid feature and to keep the downgrade experience smoother, this change set is added. 

#### PR fixes following issue(s)
Fixes #26877

#### Type of change
- New feature (non-breaking change which adds functionality)

#### How Has This Been Tested?
- [ ] Manual
- [ ] JUnit

#### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/ce/GitPrivateRepoHelperCEImpl.java [hunk 1]
```diff
@@ -40,4 +41,9 @@ public Mono<Boolean> isRepoLimitReached(String workspaceId, Boolean isClearCache
                             });
                 });
     }
+
+    @Override
+    public Mono<Boolean> isProtectedBranch(String branchName, GitApplicationMetadata gitApplicationMetadata) {
```
Comment (nayan-rafiq) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/ce/GitPrivateRepoHelperCEImpl.java:46]: This method does not fit with the name of the class. I think we should move this to the `GitServiceCE`

**Review by nayan-rafiq — CHANGES_REQUESTED** [review]

File: app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/ce/GitPrivateRepoHelperCEImpl.java (hunk 1 above)
Comment (AnaghHegde) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/ce/GitPrivateRepoHelperCEImpl.java:46]: @nayan-rafiq What do you mean by does not fit with the class? This is a CE fall back version when the feature flag is false.

Comment (nayan-rafiq) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/ce/GitPrivateRepoHelperCEImpl.java:46]: The name of the class is PrivateRepoHelper but it's dealing with the protected branch logic. Hence I'm thinking that we can move the `isProtectedBranch` inside the GitServiceCE or rename this class to something like GitHelper.

Comment (AnaghHegde) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/ce/GitPrivateRepoHelperCEImpl.java:46]: Okay. But these are all paid features and hence grouping them in a single place would be helpful. I will rename this GitHelper.

**Review by nayan-rafiq — APPROVED** [review]

**Pull request merged**